src/random.h

```diff
@@ -67,7 +67,17 @@
  * Thread-safe.
  */
 void GetRandBytes(unsigned char* buf, int num) noexcept;
+/** Generate a random integer in the range [0..range). Precondition: range > 0 */
 uint64_t GetRand(uint64_t nMax) noexcept;
+// The following GetRand*(std::chrono::*) MUST NOT be templated on the type of
+// the duration to prevent unsafe use at the call site. If it was templated, it
+// would be impossible for the compiler to select the right template. E.g.
+// assuming that a GetRand<T> exists for all duration types, then `foo` will
```

Comment:
I don't understand this.

Reply:
Oh you mean because it would use the type of the input (hours) as precision, effectively, and then converted to microseconds afterwards?

That's a very roundabout way of describing the problem, and wouldn't apply to a `GetRandMicros` that always interpreted the input as microseconds.

Reply:
Here is the code. I am happy to remove the sentence and replace it with code instead.

```cpp
template <typename D>
D GetRandDur(const D& duration_max)
{
    return D{GetRand(duration_max.count())};
}

BOOST_AUTO_TEST_CASE(util_time_GetRandTime)
{
    std::chrono::seconds rand_hour = GetRandDur(std::chrono::hours{1});
    // Want seconds to be in range [0..1hour), but always get zero :((((
    BOOST_CHECK_EQUAL(rand_hour.count(), 0);
}

Reply:
> That's a very roundabout way of describing the problem, and wouldn't apply to a GetRandMicros that always interpreted the input as microseconds.

Yes, that is the reason why I added GetRandMicros in fa883ab35ad2d4328e35b1e855d0833740a6b910, but I forgot to add documentation back then.

Reply:
Yeah, I understand now. You mean templatizing to avoid the "Micros" in the function name (and implicitly pass the resolution of the result that).

There's a rather crazy way to work around that:

```c++
template<typename In, typename OutRep = typename In::rep, typename OutPeriod = typename In::period>
std::chrono::duration<OutRep, OutPeriod> GetRand(std::chrono::duration<OutRep, OutPeriod> arg) {
    return arg;
}
```
That's a GetRand that only applies to durations, but you still always have to explicitly state the resolution you want (e.g. `GetRand<std::chrono::milliseconds>(SOME_CONSTANT_OF_TYPE_CHRONO_HOURS)` will work as expected; `GetRand(...)` won't compile).

Reply:
I doubt that's more appealing than just GetRandMicros, GetRandMillis, ... functions, though.